thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -2928,12 +3063,35 @@ class HashDedupSlaveActivityBase : public CSlaveActivity
                 lastEog = false;
                 if (bucketHandler->addRow(row)) // true if new, i.e. non-duplicate (does not take ownership)
                 {
-                    dataLinkIncrement();
-                    return row.getClear();
+                    if (!keepBest)
+                    {
+                        dataLinkIncrement();
+                        return row.getClear();
+                    }
                 }
             }
             else
             {
+                if (keepBest && !keepBestRowsReadyToReturn)
+                {
+                    bucketHandler->clearCallbacks();
+                    // This should cause one of the buckets to spill (used for testing)
+                    if (testSpillTimes)
+                    {
+                        bucketHandler->spillBucket(false);
+                        testSpillTimes--;
+                    }
+                    bestRowStream.setown(bucketHandler->getRowStream());
```

Comment:
As discussed, the spilling should be active so that the buckets can still be spilt. 

At this point (at end of stream), you could mark all non-spilt buckets as 'final', i.e. ready to stream.
+ They can still be spilt, but should not be reprocessed by the getNextBucketHandler code that handles the buckets that spilt before finished.
+ Ask the bucketHandler for next bucket stream, which may be one that's unspilt, or a previously spilt bucket that's marked as 'final'
NB: you probably don't need to clear and reinstate the memory callbacks here, which introduces a window where nothing to spill if OOM, you can instead block during this process, better to block the callback with a crit.